src/solvers/dg_triangles/dg.jl

```diff
@@ -0,0 +1,287 @@
+end
+
+# terminate recursion
+calc_boundary_flux!(cache, t, boundary_conditions::NamedTuple{(),Tuple{}}, 
+                    mesh, equations, dg::DG{<:RefElemData}) = nothing
+
+function calc_single_boundary_flux!(cache, t, boundary_condition, boundary_key, 
+                                    mesh, equations, dg::DG{<:RefElemData{Dim}}) where {Dim}
+  
+  rd = dg.basis
+  md = mesh.md
+  @unpack u_face_values, flux_face_values = cache  
+  @unpack xyzf, nxyzJ, sJ = md
+  @unpack surface_flux = dg.surface_integral
+
+  # reshape face/normal arrays to have size = (num_points_on_face, num_faces_total).
+  # mesh.boundary_faces indexes into the columns of these face-reshaped arrays.
+  num_pts_per_face = rd.Nfq ÷ rd.Nfaces
+  num_faces_total = rd.Nfaces * md.num_elements   
+  reshape_by_face(u) = reshape(u, num_pts_per_face, num_faces_total)
+  u_face_values = reshape_by_face(u_face_values)
+  flux_face_values = reshape_by_face(flux_face_values)
+  sJ      = reshape_by_face(sJ)
+  nxyzJ, xyzf   = reshape_by_face.(nxyzJ), reshape_by_face.(xyzf) # broadcast over nxyzJ::NTuple{Dim,Matrix}
+    
+  # loop through boundary faces, which correspond to columns of reshaped u_face_values, ...
+  for f in mesh.boundary_faces[boundary_key]
+    for i in Base.OneTo(num_pts_per_face)      
+      face_normal = SVector{Dim}(getindex.(nxyzJ, i, f)) / sJ[i,f]
+      face_coordinates = SVector{Dim}(getindex.(xyzf, i, f))
+      flux_face_values[i,f] = boundary_condition(u_face_values[i,f], 
+                          face_normal, face_coordinates, t,
+                          surface_flux, equations) * sJ[i,f]
+    end
+  end
+  
+  # Note: modifying the values of the reshaped array modifies the values of cache.flux_face_values. 
+  # However, we don't have to re-reshape, since cache.flux_face_values still retains its original shape.
+end
+   
+# todo: specialize for modal DG on curved meshes using WADG
+function invert_jacobian!(du, mesh::Mesh, equations, dg::DG{<:RefElemData}, 
+                          cache) where {Mesh <: AbstractMeshData} 
+  for i in Base.OneTo(ndofs(mesh,dg,cache))
+    du[i] *= -cache.invJ[i]
+  end
+end
+
+calc_sources!(du, u, t, source_terms::Nothing, 
+              mesh::VertexMappedMesh, equations, dg::DG{<:RefElemData}, cache) = nothing
+
+# uses quadrature + projection to compute source terms. 
+# todo: use interpolation here instead of quadrature? Would be cheaper. 
+function calc_sources!(du, u, t, source_terms, 
+                       mesh::VertexMappedMesh, equations, dg::DG{<:RefElemData}, cache) 
+
+  rd = dg.basis
+  md = mesh.md
+  @unpack Pq = rd
+  @unpack u_values, flux_values = cache
+  for e in Base.OneTo(md.num_elements)
+    u_e = view(u_values, :, e) # u_values should already be computed from volume kernel
+
+    # todo: why do these lines allocate?
+    for i in Base.OneTo(rd.Nq)
+      flux_values[i] = source_terms(u_e[i], getindex.(md.xyzq, i, e), t, equations) 
+    end
```

Comment:
What is returned by `getindex.(md.xyzq, i, e)`?

Reply:
`md.xyzq` is a `Tuple{Matrix{Tv}} where {Tv}` so it returns a tuple consisting of `xq[i,e], yq[i,e]`, where `xq, yq` are coordinates of quadrature points.

Reply:
I had tried converting it to an `SVector` and inlining `source_terms`, but they didn't seem to make much difference.

Reply:
I will have a look at this part soon. How did you notice that it allocates (which elixir did you run)?

Reply:
Thanks! I think any of the Euler elixirs should reproduce this, but I noticed this from the `@trixi_timeit` output for `examples/triangular_mesh_2D/elixir_euler_periodic_triangular_mesh.jl`.

Reply:
`rhs!` is not specialized on `source_terms`. Thus, there is dynamic dispatch when calling `calc_sources!`. You can avoid that by forcing specialization of `rhs!` on `source_terms` as in 
```julia
function rhs!(du, u, t, mesh, equations,
              initial_condition, boundary_conditions, source_terms::SourceTerms,
              dg::DG{<:RefElemData}, cache) where {SourceTerms}
```
However, that should only be a minor performance improvement (if significant at all), since this dynamic dispatch is only happening once per call of `rhs!`.